File: galaxy-workflow/tools/create-opends/create_opends.rb
require 'json'
require 'optparse'

options = {}
OptionParser.new do |opts|
  opts.on("--catalog_number=CATALOG_NUMBER", "Catalog number") do |catalog_number|
    options[:catalog_number] = catalog_number
  end
    opts.on("--image_license=IMAGE_LICENSE", "Image license") do |image_license|
    options[:image_license] = image_license
  end
    opts.on('--image_uri=IMAGE_URI', "Image URI") do |image_uri|
    options[:image_uri] = image_uri
  end
    opts.on("--object_type=OBJECT_TYPE", "Object Type") do |object_type|
    options[:object_type] = object_type
  end
    opts.on("--rights_holder=RIGHTS_HOLDER", "Rights Holder") do |rights_holder|
    options[:rights_holder] = rights_holder
  end
    opts.on("--registered_institution_url=INSTITUTIONAL_URL", "Registered Institution URL number") do |registered_institution_url|
    options[:registered_institution_url] = registered_institution_url
  end
    opts.on("--higher_classification=HIGHER_CLASSIFICATION", "Higher Classification") do |higher_classification|
    options[:higher_classification] = higher_classification
  end
    opts.on("--person_name[=PERSON_NAME]", "Person Name") do |person_name|
    options[:person_name] = person_name
  end
    opts.on("--person_identifier[=PERSON_IDENTIFIER]", "Person Identifier") do |person_identifier|
    options[:person_identifier] = person_identifier
  end
end.parse!

open_DS = Hash.new

open_DS[:authoritative] =
  { :physicalSpecimenId => options[:catalog_number],
    :institution => [
      options[:rights_holder],
      options[:registered_institution_url]
    ],
    :materialType => options[:object_type] }
open_DS[:images] = {
  :availableImages => [{
                         :source => options[:image_uri],
                         :license => options[:image_license] }]
}
open_DS[:higher_classification] = options[:higher_classification]
if(!options[:person_name].nil?)
  open_DS[:person_name] = options[:person_name]
end
if(!options[:person_identifier].nil?)
  open_DS[:person_identifier] = options[:person_identifier]
end

puts open_DS.to_json

File: galaxy-workflow/tools/create-opends/create_opends.py
import argparse
import hashlib
import json
import os
from PIL import Image
import ssl
import sys
from urllib import request


def main(args):
    ###########################################################################
    #  Populate openDS

    open_DS = {}

    open_DS['authoritative'] = {'physicalSpecimenId': args.catalog_number,
                                'institution': [args.rights_holder,
                                                args.registered_institution_url
                                                ],
                                'materialType': args.object_type
                                }

    open_DS['images'] = {'availableImages': [{'source': args.image_uri,
                                              'license': args.image_license}
                                             ]
                         }

    open_DS['higher_classification'] = args.higher_classification

    open_DS['person_name'] = args.person_name
    open_DS['person_identifier'] = args.person_identifier

    # -------------------------------------------------------------------------
    #  Write image to file

    # Create SSL context
    ssl._create_default_https_context = ssl._create_unverified_context

    # Get hashed filename
    hash_object = hashlib.sha1(args.image_uri.encode())
    filename_hash = hash_object.hexdigest()

    # Retrieve the URI
    (filename, headers) = request.urlretrieve(args.image_uri,
                                              filename=f'/srv/galaxy/images/{filename_hash}')

    # -------------------------------------------------------------------------
    #  Populate metadata

    image = Image.open(filename)
    open_DS['payloads'] = {
        'name': 'original image',
        'filename': filename_hash,
        'width': image.width,
        'height': image.height,
        'mediaType': headers.get_content_type(),
        'size n': os.stat(filename).st_size
    }

    # -------------------------------------------------------------------------
    #  Print

    open_DS_json = json.dumps(open_DS)
    print(open_DS_json)


def parse_args(raw_args):
    parser = argparse.ArgumentParser(prog='Create OpenDS',
                                     description='Options for the create_opends wrapper')
    parser.add_argument('--catalog_number'             , required=True)
    parser.add_argument('--image_license'              , required=True)
    parser.add_argument('--image_uri'                  , required=True)
    parser.add_argument('--object_type'                , required=True)
    parser.add_argument('--rights_holder'              , required=True)
    parser.add_argument('--registered_institution_url' , required=True)
    parser.add_argument('--higher_classification'      , required=True)
    parser.add_argument('--person_name'                , required=False, default='')
    parser.add_argument('--person_identifier'          , required=False, default='')

    args = parser.parse_args(raw_args)
    return args


if __name__ == "__main__":
    args = parse_args(sys.argv[1:])
    main(args)
